### DISCUSSION TITLE
server compiled with TypeScript causes client to fail to connect with `MatchMakeError: Class constructor Room cannot be invoked without 'new'`

### DISCUSSION THREAD

**Discussion by SIGSTACKFAULT** [colyseus/colyseus#559] [Q&A]
I'm unsure if this bug is in the client or the server, honestly. the error seems to happen in colyseus.js but it's caused by a change in the server.

## Steps to Reproduce
to isolate variables, I wrote this docker container: https://gitlab.com/SIGSTACKFAULT/colesys-bug-container.  
tl;dr:
- write a simple server in Javascript
- `client.joinOrCreate()` -> works
- copy/paste the server to a typescript file
- compile with `tsc`
- `client.joinOrCreate()` -> error:
```
/root/client.js connecting to ws://localhost:2568
/root/node_modules/colyseus.js/build/cjs/Client.js:16
        var _this = _super.call(this, message) || this;
                           ^

MatchMakeError: Class constructor Room cannot be invoked without 'new'
    at new MatchMakeError (/root/node_modules/colyseus.js/build/cjs/Client.js:16:28)
    at Client.<anonymous> (/root/node_modules/colyseus.js/build/cjs/Client.js:152:35)
    at step (/root/node_modules/tslib/tslib.js:193:27)
    at Object.next (/root/node_modules/tslib/tslib.js:174:57)
    at fulfilled (/root/node_modules/tslib/tslib.js:164:62)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
  code: 4213
}

Node.js v19.8.1
```

## Context
> How has this issue affected you? What are you trying to accomplish?  

I would expect the TypeScript example code to, y'know, work.

## Your Environment

* Colyseus Version: 0.14.24
* Node.js Version: 19.6.0
* Operating System and version: Manjaro rolling
* Link to your project: [server](https://gitlab.com/SIGSTACKFAULT/fleetwright-web-designer/-/blob/netcode/vtt-server/server.ts), [client](https://gitlab.com/SIGSTACKFAULT/fleetwright-web-designer/-/blob/netcode/src/components/VTT.vue)

**Comment by endel** [comment]
Hi @SIGSTACKFAULT, this error is usually due to `compilerOptions.target` config on TypeScript: https://docs.colyseus.io/colyseus/faq/#im-getting-this-error-class-constructor-room-cannot-be-invoked-without-new-what-should-i-do

The `gameServer.define("xxx", RoomClass)` must have a valid `colyseus.Room` class reference. I couldn't find the server files on your project's link.

**Reply by SIGSTACKFAULT** [reply]
> I couldn't find the server files on your project's link.

it's on a branch *and* it turns out i haven't pushed recently. updated the OP with proper links.

**Reply by SIGSTACKFAULT** [reply]
\> set server tsconfig `compilerOptions.target` to `esnext`
\> client works

why the *frack* does the server's compile options affect the client?
Feature Request: catch this specific error and print a more helpful error message
